core/integration/client_test.go

```diff
@@ -13,3 +20,87 @@ func TestClientClose(t *testing.T) {
 	err := c.Close()
 	require.NoError(t, err)
 }
+
+func TestClientMultiSameTrace(t *testing.T) {
```

Comment:
This test covers quite a bit, and I confirmed that it fails before this PR's changes, but it's pretty sensitive. I've had it pass 100 times in a row locally but theoretically there's still a timing issue lurking with logs (see FIXME below). The timing issue is real, but I suspect it mostly affects nesting situations like our test suite.